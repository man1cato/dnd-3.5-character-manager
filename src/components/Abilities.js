import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
