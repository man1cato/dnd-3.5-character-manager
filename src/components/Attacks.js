import React from 'react';

const Attacks = (props) => (
    <div>
        <div className="row">Base Attack Bonus: {props.bab}</div>
        <div className=" grid grid--6col">
            <h4 className="grid__left-col">Attack Type</h4>
            <h4>Base</h4>
            <div></div>
            <h4>Temp Mod</h4>
            <div></div>
            <h4>Total</h4>

            <div className="grid__left-col">Melee</div>
            <div>{props.melee.base}</div>
            <div>+</div>
            <input
                type="text"
                name="melee"
                value={props.melee.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.melee.total}</div>

            <div className="grid__left-col">Ranged</div>
            <div>{props.ranged.base}</div>
            <div>+</div>
            <input
                type="text"
                name="ranged"
                value={props.ranged.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.ranged.total}</div>

            <div className="grid__left-col">Grapple</div>
            <div>{props.grapple.base}</div>
            <div>+</div>
            <input
                type="text"
                name="grapple"
                value={props.grapple.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.grapple.total}</div>
        </div>
    </div>
)

export default Attacks