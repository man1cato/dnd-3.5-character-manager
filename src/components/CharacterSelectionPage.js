import React from 'react'
import {connect} from 'react-redux';

import Header from './Header'
import {startSetProfile} from '../actions/profile'
import {history} from '../routers/AppRouter'


export class CharacterSelectionPage extends React.Component {
   state = {
      disabled: false 
   }

   handleClick = (e) => {
		const id = e.currentTarget.id
		this.setState(() => ({
			disabled: true
		}))
      this.props.startSetProfile(id)
      setTimeout(() => { history.push('/profile') }, 1500)
   }

   render() {
      return (
         <div className="layout">
            <Header pageTitle="Character Selection" />

            <div className="container container--body">
               <h3>Select your character:</h3>
               {this.props.profiles.map((profile, i) => (

                  <button 
                     key={i}
                     className="button-profile"
                     id={profile.id}   
                     disabled={this.state.disabled}
                     onClick={this.handleClick}
                  >
                     <img src={profile.iconUrl} />
                     <div>
                        <h3>{profile.name}</h3>
                        <div>{profile.gender} {profile.race}</div>
                        <div>{profile.jobClass}</div>
                        <div>{profile.alignment}</div>
                        <div>{`Level ${profile.level}`}</div>
                     </div>
                  </button>
               ))}

            </div>
         </div>
      )
   }
}

const mapStateToProps = (state) => ({
	profiles: state.profiles
})

const mapDispatchToProps = (dispatch, props) => ({
   startSetProfile: (id) => dispatch(startSetProfile(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(CharacterSelectionPage)