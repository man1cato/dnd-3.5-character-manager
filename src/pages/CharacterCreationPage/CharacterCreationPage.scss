@import '../../styles/base/settings';


.CreatorForm {
   height: 100vh;
   display: flex;
   flex-direction: column;
   overflow: hidden;
}

.CreatorForm__group {
   display: flex;
   justify-content: space-between;
   align-items: center;
   margin-bottom: $m-size;
}

.CreatorForm__input-group {
   display: flex;
   align-items: center;
   > * {
      margin-right: $s-size;
   }
   input {
      max-width: 70px; 
   }
   > :last-child {
      margin-right: 0;
   }
}

.CreatorForm__grid {
	display: grid;
   grid-template-rows: 2rem auto;
	grid-gap: $xs-size;
	justify-items: center;
	align-items: center;
	text-align: center; 
   font-size: $font-size-m;
}

.CreatorForm__error {
   font-style: italic;
   font-size: $font-size-s;
   margin: 0;
}

.CreatorForm__error--center {
   @extend .CreatorForm__error;
   text-align: center;
}

.CreatorForm__error--right {
   @extend .CreatorForm__error;
   text-align: right;
}