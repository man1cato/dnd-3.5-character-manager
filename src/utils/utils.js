import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])