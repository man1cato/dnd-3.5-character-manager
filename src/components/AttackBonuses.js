import React, { useState, useEffect, Fragment } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { calcBaseMeleeBonus, calcBaseRangedBonus, calcBaseGrappleBonus, convertInputValue } from '../utils/utils'


const AttackBonuses = ({ profileAttackBonusMods, abilities, baseAttackBonuses, size, handleUpdate }) => {  
    const attackBonusBases = {
        melee: _.map(baseAttackBonuses, (bab) => calcBaseMeleeBonus(bab, abilities.str.score, size)),
        ranged: _.map(baseAttackBonuses, (bab) => calcBaseRangedBonus(bab, abilities.dex.score, size)),
        grapple: _.map(baseAttackBonuses, (bab) => calcBaseGrappleBonus(bab, abilities.str.score, size))
    }
    const [attackBonusMods, setAttackBonusMods] = useState(_.mapValues(profileAttackBonusMods, (val) => val || '')) 
    const [attackBonusTotals, setAttackBonusTotals] = useState(
        _.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val ))
    ) 
    
    useEffect(() => {
        setAttackBonusTotals(_.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val)))
        handleUpdate({ attackBonusMods })
    }, [attackBonusMods])

    return (    
        <div className="grid--6col">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(['melee', 'ranged', 'grapple'], (key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{attackBonusBases[key].join(' / ')}</div>
                    <div>+</div>
                    <input
                        type="number"
                        id={key}
                        value={attackBonusMods[key]}
                        onChange={(e) => {
                            setAttackBonusMods(
                                update(attackBonusMods, {
                                    [key]: { $set: convertInputValue(e.target.value) }
                                })
                            )
                        }}
                    />
                    <div>=</div>
                    <div>{attackBonusTotals[key].join(' / ')}</div>
                </Fragment>
            ))}

        </div>
    )
}


export default AttackBonuses