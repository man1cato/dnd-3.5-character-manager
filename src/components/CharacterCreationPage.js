import React from 'react'
import {connect} from 'react-redux'
import Header from './Header'
import {Formik, Form} from 'formik'
import * as Yup from 'yup'

import Page1 from './CreatorFormPage1'
import Page2 from './CreatorFormPage2'
import Page3 from './CreatorFormPage3'
import CreatorFormFooter from './CreatorFormFooter'
import {history} from '../routers/AppRouter'
import {startCreateProfile} from '../actions/profile'
import {getSpecialAbilityIdsFromLevels} from '../utils/getSpecialAbilities'


const pages = [Page1, Page2, Page3]

const validationSchema = Yup.object().shape({
	page1: Yup.object().shape({
		name: Yup.string().min(2, 'Name is too short!').max(30, 'Name is too long!').required('Required'),
		age: Yup.number().required().positive().integer(),
		heightFt: Yup.number('Enter a positive integer').positive('Enter a positive integer').integer('Enter a positive integer').required('Required'),
		heightIn: Yup.number('Enter an integer between 0 and 12').positive('Enter an integer between 0 and 12').integer('Enter an integer between 0 and 12').max(11, 'Enter an integer between 0 and 12').required('Required'),
		weight: Yup.number('Enter a positive integer').positive('Enter a positive integer').integer('Enter a positive integer').required('Required')
	}),
	page2: undefined,
	page3: undefined
})

export class CharacterCreationPage extends React.Component {
	state = {
		selectedRace: this.props.races.find((race) => race.name === 'Human'),
		selectedJobClass: this.props.jobClasses.find((jobClass) => jobClass.name === 'Fighter'),
		page: 1
	}

	handleSelect = (e, setFieldValue, setFieldError) => {
		const name = e.target.name;
		const value = e.target.value;
		if (name === 'race') { setFieldValue('bonusLanguages', []) } 
		if (name === 'school') {
			setFieldValue('prohibitedSchools', []);
			if (value === 'Universal') {
				setFieldError('prohibitedSchools', undefined);
			} 
		} 
		if (name === 'jobClass') {
			if (value === this.props.jobClasses.find((jobClass) => jobClass.name === 'Paladin').id) { 
				setFieldValue('alignment', 'Lawful Good') 
			}
			if (value === this.props.jobClasses.find((jobClass) => jobClass.name === 'Wizard').id) { 
				setFieldValue('school', 'Universal') 
			} else {
				setFieldValue('school', null)
			}
		}
		
		this.setState((prevState) => {
			const selectedRace = name === 'race' ? this.props.races.find((race) => race.id === value) : prevState.selectedRace;
			const selectedJobClass = name === 'jobClass' ? this.props.jobClasses.find((jobClass) => jobClass.id === value) : prevState.selectedJobClass;
			
			return {
				selectedRace,
				selectedJobClass
			}
		});
	}

	handleMultiSelect = (e, setFieldValue) => {
		const name = e.target.name;
		const options = e.target.options;
		const value = [];
		for (let i = 0; i < options.length; i++) {
			if (options[i].selected) {
				value.push(options[i].value);
			}
		}
		setFieldValue(name, value);
	}

	handleBack = (setErrors) => {
		this.setState((prevState) => ({
			page: prevState.page - 1
		}));        
		setErrors({});
	}

	handleNext = () => {
		this.setState((prevState) => ({
			page: prevState.page + 1
		}));
	}    

	render() {        
		return (
			<div className="layout">
				<Header pageTitle="Character Creation" />
				
				<Formik
					initialValues={{
						name: '',
						gender: 'Male',
						age: '',
						heightFt: '',
						heightIn: '',
						weight: '',
						race: this.props.races.find((race) => race.name === 'Human').id,
						jobClass: this.props.jobClasses.find((jobClass) => jobClass.name === 'Fighter').id,
						alignment: 'Lawful Good',
						deity: ''
					}}
												
					validationSchema={Yup.reach(validationSchema, `page${this.state.page}`)}

					onSubmit={(values, {setErrors, setSubmitting}) => {
						getSpecialAbilityIdsFromLevels(values.jobClass, 1).then((specialAbilities) => {
							const profile = {
								...values,
								race: this.state.selectedRace.name,
								size: this.state.selectedRace.size,
								specialAbilities,
								jobClass: this.state.selectedJobClass.name,
								height: `${values.heightFt}'${values.heightIn}"`,
								languages: this.state.selectedRace.defaultLanguages.concat(values.bonusLanguages),
								deity: !!values.deity ? values.deity : "None",
								level: 1,
								iconUrl: this.state.selectedRace.iconUrl
							}
							delete profile.heightFt
							delete profile.heightIn
							if(profile.school === null) {delete profile.school}
							this.props.startCreateProfile(profile)
						})							
						setTimeout(() => { history.push('/profile') }, 2000)
						setSubmitting(false)
					}}
				>

					{({values, setFieldValue, handleChange, isSubmitting, isValid, setErrors, setFieldError}) => (
						<Form >
							<div className="container container--body">
								{{
									1: <Page1 
										values={values}
										races={this.props.races}
										jobClasses={this.props.jobClasses}
										selectedRace={this.state.selectedRace} 
										handleChange={handleChange} 
										handleSelect={this.handleSelect}  
										setFieldValue={setFieldValue}
									/>,
									2: <Page2
										values={values}
										selectedRace={this.state.selectedRace} 
										jobClasses={this.props.jobClasses}
										selectedJobClass={this.state.selectedJobClass}
										alignments={this.props.alignments}
										schools={this.props.schools}
										handleChange={handleChange}
										handleSelect={this.handleSelect}
										handleMultiSelect={this.handleMultiSelect}
										setFieldValue={setFieldValue}
										setFieldError={setFieldError}
									/>,
									3: <Page3
										values={values}
										feats={this.props.feats}
										selectedJobClass={this.state.selectedJobClass}										
										handleChange={handleChange}
										handleSelect={this.handleSelect}
										handleMultiSelect={this.handleMultiSelect}
										setFieldValue={setFieldValue}
										setFieldError={setFieldError}
									/>
								}[this.state.page]}
							</div>
							
							<CreatorFormFooter 
								page={this.state.page}
								pages={pages}
								handleBack={this.handleBack}
								handleNext={this.handleNext}
								handleSubmit={this.handleSubmit}
								setErrors={setErrors}
								isSubmitting={isSubmitting}
								isValid={isValid}
							/>

						</Form>
					)}
				</Formik>
					
			</div>
		)
	}
};


const mapStateToProps = (state) => ({
	races: state.races,
	jobClasses: state.jobClasses,
	feats: state.feats,
	schools: [
		'Abjuration','Clairsentience', 'Conjuration', 'Divination', 
		'Enchantment', 'Evocation', 'Illusion', 'Metacreativity', 
		'Necromancy', 'Psychokinesis', 'Psychometabolism', 'Psychoportation', 
		'Telepathy', 'Transmutation', 'Universal'
	],
	alignments: [
		'Lawful Good', 'Neutral Good', 'Chaotic Good', 
		'Lawful Neutral', 'True Neutral', 'Chaotic Neutral', 
		'Lawful Evil', 'Neutral Evil', 'Chaotic Evil'
	]
});

const mapDispatchToProps = (dispatch, props) => ({
	startCreateProfile: (profile) => dispatch(startCreateProfile(profile))
});


export default connect(mapStateToProps, mapDispatchToProps)(CharacterCreationPage);
