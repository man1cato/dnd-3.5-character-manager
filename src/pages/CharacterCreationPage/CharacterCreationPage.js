import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Formik, Form } from 'formik'
import * as Yup from 'yup'
import _ from 'lodash'

import { history } from '../../routers/AppRouter'
import { startCreateProfile } from '../../store/actions/profile'
import { apiObjectToArray, calcAbilityMod, convertMoneyToDenominations, calcAttackBonuses } from '../../utils/utils'
import { abilities, sizeMods } from '../../utils/staticData'

import CreatorFormIdentity from './CreatorFormIdentity/CreatorFormIdentity'
import CreatorFormJobClass from './CreatorFormJobClass/CreatorFormJobClass'
import CreatorFormAbilities from './CreatorFormAbilities/CreatorFormAbilities'
import CreatorFormFeats from './CreatorFormFeats/CreatorFormFeats'
import CreatorFormSkills from './CreatorFormSkills/CreatorFormSkills'
import CreatorFormEquipment from './CreatorFormEquipment/CreatorFormEquipment'
import CreatorFormFooter from './CreatorFormFooter/CreatorFormFooter'


const pages = [CreatorFormIdentity, CreatorFormJobClass, CreatorFormAbilities, CreatorFormFeats, CreatorFormSkills, CreatorFormEquipment]

const abilityMessage = 'Enter an integer between 3 and 18'
const abilityValidation = Yup.number().integer(abilityMessage).min(3, abilityMessage).max(18, abilityMessage).required('Required').typeError(abilityMessage)
const validationSchema = Yup.object().shape({
	page1: Yup.object().shape({
		name: Yup.string().max(30, 'Name is too long!').required('Required').typeError('Required'),
		age: Yup.number().positive().integer().required('Required').typeError('Enter a number'),
		height: Yup.object().shape({
			ft: Yup.number().positive('Enter a positive integer').integer('Enter a positive integer').required('Required').typeError('Enter a number').typeError('Enter a number'),
			in: Yup.number().positive('Enter an integer between 0 and 12').integer('Enter an integer between 0 and 12').max(11, 'Enter an integer between 0 and 12').required('Required').typeError('Enter a number'),
		}),
		weight: Yup.number().positive('Enter a positive integer').integer('Enter a positive integer').required('Required').typeError('Enter a number')
	}),
	page2: Yup.object().shape({
		startingGold: Yup.number().positive('Cannot be negative').integer('Enter an integer')
	}),
	page3: Yup.object().shape({
		abilities: Yup.object().shape(
			_.mapValues(abilities, () => Yup.object().shape({ score: abilityValidation }))
		)
	}),
	page4: Yup.object().shape({
		feats: Yup.array().required('Select at least one feat')
	}),
	page5: Yup.object().shape({
		remainingSkillPoints: Yup.number().moreThan(-1, 'Too many skill points assigned').lessThan(1, 'Assign all skill points')
	}),
	page6: Yup.object().shape({
		equipment: Yup.array().required('Select at least one equipment'),
		remainingGold: Yup.number().moreThan(-1, 'Purchases must not exceed available money')
	})
})


export const CharacterCreationPage = props => {
	const [page, setPage] = useState(1)

	const handleMultiSelect = (e, setFieldValue) => {
		const name = e.target.name
		const options = e.target.options
		const value = []
		for (let i = 0; i < options.length; i++) {
			if (options[i].selected) {
				value.push(options[i].value)
			}
		}
		setFieldValue(name, value)
	}
	
	return (
		<Formik
			initialValues={{
				name: '',
				gender: 'Male',
				age: '',
				height: {
					ft: '',
					in: ''
				},
				weight: '',
				alignment: 'Lawful Good',
				race: _.find(props.races, race => race.name === 'Human'),
				jobClass: _.find(props.jobClasses, jobClass => jobClass.name === 'Fighter'),
				bonusLanguages: [],
				deity: '',
				abilities: _.mapValues(abilities, () => ({ score: '', final: '' })),
				feats: [],
				equipment: [],
				equipped: {
					armor: null,
					shield: null,
					weapons: []
				},
				skillSet: apiObjectToArray(props.skills).map(skill => ({id: skill.id, ranks: 0 })),
				skillPoints: 0,
				remainingSkillPoints: 0,
				startingGold: 0,
				remainingGold: 0
			}}
										
			validationSchema={Yup.reach(validationSchema, `page${page}`)}

			onSubmit={(values, { setSubmitting }) => {
				const { 
					name, gender, age, height, weight, race, alignment, jobClass, 
					remainingGold, bonusLanguages, deity, feats, skillSet,
					equipment, equipped, school, prohibitedSchools  
				} = values

				const abilities = _.mapValues(values.abilities, ability => ({ 
					score: ability.final,
					tempScore: ''
				}))
				const dexMod = calcAbilityMod(abilities.dex.score)
				let armorBonus = 0
				if (equipped.armor) { armorBonus += props.items[equipped.armor].armorBonus }
				if (equipped.shield) { armorBonus += props.items[equipped.shield].armorBonus }
				const baseArmorClass = 10 + armorBonus + sizeMods[race.size].size + dexMod
				
				const profile = {
					name,
					gender,
					age,
					height: `${height.ft}'${height.in}"`,
					weight,
					race: _.findKey(props.races, race),
					alignment,
					jobClass: _.findKey(props.jobClasses, jobClass),
					money: convertMoneyToDenominations(remainingGold),
					languages: _.sortBy(race.defaultLanguages.concat(bonusLanguages)),
					specialAbilities: jobClass.levels["1"].specialAbilities,
					deity: !!deity ? deity : "None",
					abilities,
					feats,
					skillSet: _.filter(skillSet, skill => skill.ranks > 0),
					equipment,
					equipped,
					ac: {
						base: baseArmorClass,
						flat: baseArmorClass - dexMod,
						touch: baseArmorClass - armorBonus
					},
					attackBonuses: _.mapValues(
						calcAttackBonuses(
							jobClass.levels[1].baseAttackBonuses || [0],
							race.size,
							abilities.str.score,
							abilities.dex.score
						),
						val => ({ base: val, mod: 0, total: val })
					),
					level: 1,
					hp: {
						base: Number(jobClass.hitDie.slice(1)) + calcAbilityMod(abilities.con.score)
					},
					xp: 0,
					iconUrl: race.iconUrl
				}
				if(!!school) {
					profile.school = school 
					profile.prohibitedSchools = prohibitedSchools
				}
				console.log('Created profile: ', profile)
				props.startCreateProfile(profile)
				
				setTimeout(() => { history.push('/profile') }, 1500)
				setSubmitting(true)
			}}
		>
			{({ values, setFieldValue, handleChange, handleSubmit, isSubmitting, isValid, validateForm, setErrors, setFieldError, setTouched}) => (
				<Form className="CreatorForm">
					{{
						1: <CreatorFormIdentity 
							values={values}
							races={props.races}
							jobClasses={props.jobClasses}
							setFieldValue={setFieldValue}
							setTouched={setTouched}
						/>,
						2: <CreatorFormJobClass
							values={values}
							jobClasses={props.jobClasses}
							handleChange={handleChange}
							handleMultiSelect={handleMultiSelect}
							setFieldValue={setFieldValue}
							setFieldError={setFieldError}
							validateForm={validateForm}
						/>,
						3: <CreatorFormAbilities
							values={values}
							handleChange={handleChange}
							setFieldValue={setFieldValue}
							validateForm={validateForm}								
						/>,
						4: <CreatorFormFeats
							values={values}
							feats={props.feats}
							setFieldValue={setFieldValue}
						/>,
						5: <CreatorFormSkills
							values={values}
							skills={props.skills}
							setFieldValue={setFieldValue}
						/>,
						6: <CreatorFormEquipment
							values={values}
							items={props.items}
							setFieldValue={setFieldValue}
						/>
					}[page]}

					<CreatorFormFooter 
						page={page}
						pageCount={pages.length}
						setPage={setPage}
						handleSubmit={handleSubmit}
						setErrors={setErrors}
						isSubmitting={isSubmitting}
						isValid={isValid}
						validateForm={validateForm}
					/>

				</Form>
			)}
		</Formik>				
	)
}


const mapStateToProps = state => ({
	races: state.races,
	jobClasses: state.jobClasses,
	feats: _.omitBy(state.feats, feat => feat.prerequisites || _.includes(feat.types, 'Epic') || _.includes(feat.types, 'Creature')),
	items: _.omitBy(state.items, item => item.weaponType === 'Natural' || _.includes(['Creature Part', 'Money'], item.category)),
	skills: state.skills
})

const mapDispatchToProps = dispatch => ({
	startCreateProfile: profile => dispatch(startCreateProfile(profile))
})


export default connect(mapStateToProps, mapDispatchToProps)(CharacterCreationPage)