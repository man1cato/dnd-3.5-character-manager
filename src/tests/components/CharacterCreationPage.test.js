import React from 'react'
import { shallow } from 'enzyme'
import fs from 'fs'

import { CharacterCreationPage } from '../../components/CharacterCreationPage'
import { apiObjectToArray } from '../../utils/utils'

let api, wrapper, props

beforeAll(async () => {
	api = await fs.promises.readFile('src/tests/fixtures/api.json')
	api = JSON.parse(api)
	props = {
		races: apiObjectToArray(api.races),
		jobClasses: apiObjectToArray(api.jobClasses),
		feats: apiObjectToArray(api.feats)
	}
})

beforeEach(() => {
	wrapper = shallow(<CharacterCreationPage {...props} />)
})


test('should render CharacterCreationPage', () => {
	expect(wrapper).toMatchSnapshot()
})
