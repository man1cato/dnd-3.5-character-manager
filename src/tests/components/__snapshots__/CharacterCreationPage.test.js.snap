// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharacterCreationPage 1`] = `
<div
  className="layout"
>
  <Connect(Header)
    pageTitle="Character Creation"
  />
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "age": "",
        "alignment": "Lawful Good",
        "deity": "",
        "gender": "Male",
        "heightFt": "",
        "heightIn": "",
        "jobClass": "reckjdfZ5f86z4Rtz",
        "name": "",
        "race": "recxyRHegDBrFYJP5",
        "weight": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
    validationSchema={
      ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": undefined,
        "_nodes": Array [
          "weight",
          "heightIn",
          "heightFt",
          "age",
          "name",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "age": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "integer": false,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "heightFt": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "integer": false,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "heightIn": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "integer": false,
              "max": true,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "name": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "max": true,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "weight": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "integer": false,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
      }
    }
  >
    <Component />
  </Formik>
</div>
`;
